src/msg/producer/writer/message.go

```diff
@@ -50,18 +49,18 @@ func newMessage() *message {
 
 // Set sets the message.
 func (m *message) Set(meta metadata, rm *producer.RefCountedMessage, initNanos int64) {
+	m.isAcked.Store(false)
+	m.rm.Store(unsafe.Pointer(rm))
 	m.initNanos = initNanos
 	m.meta = meta
-	m.RefCountedMessage = rm
 	m.ToProto(&m.pb)
 }
 
 // Close resets the states of the message.
 func (m *message) Close() {
-	m.RefCountedMessage = nil
+	m.rm.Store(nil)
```

Comment:
NB: this is all moot and can be non-atomic if we fix the Buffer side of things and pool these.